libs/vm-modules/src/ml/model/model.cpp

```diff
@@ -89,233 +123,65 @@ Ptr<VMModel> VMModel::Constructor(VM *vm, TypeId type_id,
-  }
-
-  auto model_ptr = std::dynamic_pointer_cast<fetch::ml::model::Sequential<TensorType>>(model_);
-  if (layer->string() == "conv1d")
-  {
-    model_ptr->Add<fetch::ml::layers::Convolution1D<TensorType>>(
-        output_channels, input_channels, kernel_size, stride_size, activation_type);
-  }
-  else if (layer->string() == "conv2d")
-  {
-    model_ptr->Add<fetch::ml::layers::Convolution2D<TensorType>>(
-        output_channels, input_channels, kernel_size, stride_size, activation_type);
-  }
-  else
-  {
-    throw std::runtime_error("invalid params specified for " + layer->string() + " layer");
-  }
-}
-
 void VMModel::CompileSequential(fetch::vm::Ptr<fetch::vm::String> const &loss,
                                 fetch::vm::Ptr<fetch::vm::String> const &optimiser)
 {
-  fetch::ml::ops::LossType loss_type;
-  fetch::ml::OptimiserType optimiser_type;
-
-  if (loss->string() == "mse")
-  {
-    loss_type = fetch::ml::ops::LossType::MEAN_SQUARE_ERROR;
-  }
-  else if (loss->string() == "cel")
-  {
-    loss_type = fetch::ml::ops::LossType::CROSS_ENTROPY;
-  }
-  else if (loss->string() == "scel")
-  {
-    loss_type = fetch::ml::ops::LossType::SOFTMAX_CROSS_ENTROPY;
-  }
-  else
-  {
-    throw std::runtime_error("invalid loss function");
-  }
-
-  // dense / fully connected layer
-  if (optimiser->string() == "adagrad")
-  {
-    optimiser_type = fetch::ml::OptimiserType::ADAGRAD;
-  }
-  else if (optimiser->string() == "adam")
-  {
-    optimiser_type = fetch::ml::OptimiserType::ADAM;
-  }
-  else if (optimiser->string() == "momentum")
-  {
-    optimiser_type = fetch::ml::OptimiserType::MOMENTUM;
-  }
-  else if (optimiser->string() == "rmsprop")
-  {
-    optimiser_type = fetch::ml::OptimiserType::RMSPROP;
-  }
-  else if (optimiser->string() == "sgd")
-  {
-    optimiser_type = fetch::ml::OptimiserType::SGD;
-  }
-  else
-  {
-    throw std::runtime_error("invalid optimiser");
-  }
-
+  auto const loss_type      = ParseName(loss->string(), losses_, "loss function");
+  auto const optimiser_type = DeduceOptimizer(optimiser->string());
   model_->Compile(optimiser_type, loss_type);
 }
 
 void VMModel::CompileSimple(fetch::vm::Ptr<fetch::vm::String> const &        optimiser,
                             fetch::vm::Ptr<vm::Array<math::SizeType>> const &in_layers)
 {
-  // construct the model with the specified layers
-  auto                        n_elements = in_layers->elements.size();
-  std::vector<math::SizeType> layers(n_elements);
-  for (std::size_t i = 0; i < n_elements; ++i)
+  OptimiserType optimiser_type = DeduceOptimizer(optimiser->string());
+  if (optimiser_type != OptimiserType::ADAM)
   {
-    layers.at(i) = in_layers->elements.at(i);
+    throw std::runtime_error(
+        R"(Wrong optimiser, a "Simple" model can use only "adam", while given : )" +
+        optimiser->string());
   }
 
+  auto BuildLayersFromInput = [&in_layers]() {
```

Comment:
this is really unclear.
setting up a lambda function to return to an auto and passing to the DNNRegressor constructor hides a lot about what is going on.

If you want to functionalise this piece of code then make a static member function called BuildLayersFromInput that returns a vector<SizeType>. Invoke that function, and pass the results to the constructor when you need to.